File: platforms/android/app/src/main/java/wtf/fix1132/android/MeetingLink.java
package wtf.fix1132.android;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Turns whatever the user pasted into a Zoom web client link.
 *
 * The web client is the part that matters for 1132: it runs in the browser, so
 * it carries none of the native client's device identity. Feeding it a meeting
 * id directly is the most reliable bypass available on a phone.
 */
final class MeetingLink {

    /** Zoom meeting ids are 9 to 11 digits. */
    private static final Pattern MEETING_ID = Pattern.compile("(\\d[\\d\\s-]{7,20}\\d)");
    private static final Pattern PASSWORD = Pattern.compile("[?&]pwd=([^&\\s]+)");

    private MeetingLink() {
    }

    static final String JOIN_PAGE = "https://zoom.us/join";

    /**
     * @param input a meeting id, a zoom.us link, or a zoommtg:// link
     * @return a web client URL, or the generic join page when no id is found
     */
    static String toWebClientUrl(String input) {
        if (input == null) {
            return JOIN_PAGE;
        }
        String trimmed = input.trim();
        if (trimmed.isEmpty()) {
            return JOIN_PAGE;
        }

        String password = null;
        Matcher passwordMatch = PASSWORD.matcher(trimmed);
        if (passwordMatch.find()) {
            password = passwordMatch.group(1);
        }

        String id = extractMeetingId(trimmed);
        if (id == null) {
            // Not recognisable as a meeting, but if it is already a URL let the
            // browser deal with it rather than throwing it away.
            if (trimmed.startsWith("http://") || trimmed.startsWith("https://")) {
                return trimmed;
            }
            return JOIN_PAGE;
        }

        StringBuilder url = new StringBuilder("https://zoom.us/wc/join/").append(id);
        if (password != null && !password.isEmpty()) {
            url.append("?pwd=").append(password);
        }
        return url.toString();
    }

    /** Pulls the digits of a meeting id out of free-form text. */
    static String extractMeetingId(String input) {
        if (input == null) {
            return null;
        }
        // A /j/<id> or /wc/join/<id> path segment is the most reliable signal,
        // so look for it before falling back to any run of digits.
        Matcher pathMatch = Pattern.compile("/(?:j|wc/join|s)/(\\d{9,11})").matcher(input);
        if (pathMatch.find()) {
            return pathMatch.group(1);
        }

        Matcher loose = MEETING_ID.matcher(input);
        while (loose.find()) {
            String digits = loose.group(1).replaceAll("[^0-9]", "");
            if (digits.length() >= 9 && digits.length() <= 11) {
                return digits;
            }
        }
        return null;
    }
}

File: platforms/ios/1132WTF/MeetingLink.swift
import Foundation

/// Turns whatever the user pasted into a Zoom web client link.
///
/// The web client runs in a web view, so it carries none of the native app's
/// device identity. On iOS that makes it the only level 1132.WTF can carry out
/// entirely on its own.
enum MeetingLink {
    static let joinPage = "https://zoom.us/join"

    /// - Parameter input: a meeting id, a zoom.us link, or a zoommtg:// link.
    /// - Returns: a web client URL, or the generic join page when no id is found.
    static func webClientURL(from input: String) -> String {
        let trimmed = input.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return joinPage }

        let password = extractPassword(from: trimmed)

        guard let id = meetingID(from: trimmed) else {
            // Not recognisable as a meeting, but if it is already a URL then let
            // the web view deal with it rather than throwing it away.
            if trimmed.hasPrefix("http://") || trimmed.hasPrefix("https://") {
                return trimmed
            }
            return joinPage
        }

        var url = "https://zoom.us/wc/join/\(id)"
        if let password, !password.isEmpty {
            url += "?pwd=\(password)"
        }
        return url
    }

    /// Pulls the digits of a meeting id out of free-form text.
    static func meetingID(from input: String) -> String? {
        // A /j/<id> or /wc/join/<id> path segment is the most reliable signal, so
        // look for it before falling back to any run of digits.
        if let match = firstGroup(in: input, pattern: "/(?:j|wc/join|s)/([0-9]{9,11})") {
            return match
        }

        // Meeting ids are often written with spaces or dashes between groups.
        if let loose = firstGroup(in: input, pattern: "([0-9][0-9\\s-]{7,20}[0-9])") {
            let digits = loose.filter(\.isNumber)
            if digits.count >= 9 && digits.count <= 11 {
                return digits
            }
        }
        return nil
    }

    private static func extractPassword(from input: String) -> String? {
        firstGroup(in: input, pattern: "[?&]pwd=([^&\\s]+)")
    }

    private static func firstGroup(in input: String, pattern: String) -> String? {
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return nil }
        let range = NSRange(input.startIndex..., in: input)
        guard let match = regex.firstMatch(in: input, range: range),
              match.numberOfRanges > 1,
              let group = Range(match.range(at: 1), in: input)
        else {
            return nil
        }
        return String(input[group])
    }
}
